XML parsing is vulnerable to an XXE XSRF attack.

skington opened the pull request:
XXE: XML eXternal Entity injection
XSRF: cross-Site Request Forgery

[Explanation](https://portswigger.net/web-security/xxe).

skington committed:
Test whether XML parsing is vulnerable to an XXE XSRF attack.
XXE: XML eXternal Entity injection
XSRF: cross-Site Request Forgery

bigpresh commented:
I'm interested to know if this is only an issue for one or both of the XML::Simple backends (XML::SAX and/or XML::Parser).

I can't see any way to fix it in our calls to XML::Parser itself, but it may be fixable by controlling which backend XML::Simple can use.

Other options would include re-writing Dancer::Serializer::XML to use a better XML handling library, or deprecating Dancer::Serializer::XML - I like the latter, but wouldn't want to force such a breaking change on users if at all avoidable!

skington commented:
XML::Parser, which is what ends up getting used no matter what, has an [ExternEnt handler](https://metacpan.org/pod/XML::Parser#ExternEnt-(Expat,-Base,-Sysid,-Pubid)) which can be set to "no, don't query the public Internet for something like this".

skington committed:
Disable fetching XML external entities under XML::Parser.
This causes annoying warnings because the ability to pass options like
this to XML::Parser is officially deprecated by XML::Simple.

skington committed:
Delve deeply into the guts of XML::Parser to turn off external entity
parsing. This is seriously nasty and wrong.

skington committed:
Restore presh-style line breaks.

bigpreshkt commented on lib/Dancer/Serializer/XML.pm:
```diff
@@ -71,7 +83,18 @@ sub deserialize {
+    # but appears to only be used deep in the XML::Parser XS guts.
+    no warnings 'redefine';
+    local *XML::Parser::lwp_ext_ent_handler = sub { return };
+    local $INC{'XML/Parser/LWPExternEnt.pl'} = 'Do not load this!';
```
I'll tweak the value to be e.g. 'Dancer::Serializer::XML disabled loading this to patch around XXE vulnerabilities' or similar, just in case it saves anyone some head-scratching in future.

bigpreshkt commented:
Many thanks @skington - will get a release out ASAP with that fix.  The fix is a bit nasty, but I don't see a better way, so I think it's a necessary evil.

@skington++ !
